server-spi/src/main/java/org/keycloak/models/ClientProvider.java

```diff
@@ -20,30 +20,87 @@
+     * @param realm Realm.
+     * @return List of the clients. Never returns {@code null}.
+     */
+    default List<ClientModel> getClients(RealmModel realm) {
+        return this.getClients(realm, null, null);
+    };
 
-    ClientModel addClient(RealmModel realm, String clientId);
+    /**
+     * Adds a client with given {@code clientId} to the given realm.
+     * The internal ID of the client will be created automatically.
+     * @param realm Realm owning this client.
+     * @param clientId String that identifies the client to the external parties.
+     *   Maps to {@code client_id} in OIDC or {@code entityID} in SAML.
+     * @return Model of the created client.
+     */
+    default ClientModel addClient(RealmModel realm, String clientId) {
+        return addClient(realm, null, clientId);
+    }
 
+    /**
+     * Adds a client with given internal ID and {@code clientId} to the given realm.
+     * @param realm Realm owning this client.
+     * @param id Internal ID of the client or {@code null} if one is to be created by the underlying store
+     * @param clientId String that identifies the client to the external parties.
+     *   Maps to {@code client_id} in OIDC or {@code entityID} in SAML.
+     * @return Model of the created client.
+     * @throws IllegalArgumentException If {@code id} does not conform
+     *   the format understood by the underlying store.
+     */
     ClientModel addClient(RealmModel realm, String id, String clientId);
 
-    RoleModel addClientRole(RealmModel realm, ClientModel client, String name);
-
-    RoleModel addClientRole(RealmModel realm, ClientModel client, String id, String name);
+    /**
+     * Returns number of clients in the given realm
+     * @param realm Realm.
+     * @return Number of the clients in the given realm.
+     */
+    long getClientsCount(RealmModel realm);
 
-    RoleModel getClientRole(RealmModel realm, ClientModel client, String name);
+    List<ClientModel> getAlwaysDisplayInConsoleClients(RealmModel realm);
 
-    Set<RoleModel> getClientRoles(RealmModel realm, ClientModel client);
+    /**
+     * Removes given client from the given realm.
+     * @param id Internal ID of the client
+     * @param realm Realm.
+     * @return {@code true} if the client existed and has been removed, {@code false} otherwise.
+     * @deprecated Use {@link #removeClient(RealmModel, String)} instead.
+     */
+    default boolean removeClient(String id, RealmModel realm) { return this.removeClient(realm, id); }
 
-    List<ClientModel> getAlwaysDisplayInConsoleClients(RealmModel realm);
+    /**
+     * Removes given client from the given realm.
+     * @param realm Realm.
+     * @param id Internal ID of the client
+     * @return {@code true} if the client existed and has been removed, {@code false} otherwise.
+     */
+    default boolean removeClient(RealmModel realm, String id) { return this.removeClient(id, realm); }
```

Comment:
Do we really want to introduce `default` implementation of this method? I can imagine that someone who will implement this interface could potentially forgot to override one of `removeClient` methods and it would lead to infinite loop, wouldn't it? If I understand it correctly I'd suggest to leave `default` implementation for `removeClient(String id, RealmModel realm)` and remove `default` from this one. Correct me if I am wrong please.

Reply:
Yup, you're right, I'll remove that one